lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -29,4 +98,276 @@
 # to 0 if you are raising the major API version
 LIBPATCH = 1
 
-# TODO: add your code here! Happy coding!
+from ipaddress import IPv4Address, IPv4Network
+from copy import deepcopy
+from typing import Dict, List, Union
+from ops.framework import Object, EventSource, EventBase, ObjectEvents
+from ops.charm import CharmBase
+from ops import RelationJoinedEvent, RelationChangedEvent, RelationDepartedEvent, StoredState
+import logging, json
+
+logger = logging.getLogger(__name__)
+
+Network = Dict[
+    str,  # 'network' | 'gateway' | 'routes'
+    Union[
+        IPv4Address,  # gateway, ex: '192.168.250.1'
+        IPv4Network,  # network, ex: '192.168.250.0/24'
+        List[  # List of routes
+            Dict[
+                str,  # 'destination' | 'gateway'
+                Union[
+                    IPv4Address,  # gateway, ex: '192.168.250.3'
+                    IPv4Network,  # destination, ex: '172.250.0.0/16'
+                ],
+            ]
+        ],
+    ],
+]
+
+RoutingTable = Dict[
+    str,  # Name of the application
+    List[Network],  # All networks for this application
+]
```

Comment:
Love the type aliases! We could add the type hints to make it clear what they are:
```suggestion
from typing import TypeAlias

Network: TypeAlias = Dict[
    str,  # 'network' | 'gateway' | 'routes'
    Union[
        IPv4Address,  # gateway, ex: '192.168.250.1'
        IPv4Network,  # network, ex: '192.168.250.0/24'
        List[  # List of routes
            Dict[
                str,  # 'destination' | 'gateway'
                Union[
                    IPv4Address,  # gateway, ex: '192.168.250.3'
                    IPv4Network,  # destination, ex: '172.250.0.0/16'
                ],
            ]
        ],
    ],
]

RoutingTable: TypeAlias = Dict[
    str,  # Name of the application
    List[Network],  # All networks for this application
]
```

Reply:
Done, had to move the import above